src/kem/kyber/arithmetic.rs

```diff
@@ -2,39 +2,60 @@ use super::constants::{COEFFICIENTS_IN_RING_ELEMENT, FIELD_MODULUS};
 
 pub(crate) type KyberFieldElement = i32;
 
-const BARRETT_SHIFT: i32 = 26;
-const BARRETT_R: i32 = 1i32 << BARRETT_SHIFT;
-const BARRETT_MULTIPLIER: i32 = 20159; // floor((BARRETT_R / FIELD_MODULUS) + 0.5)
+const MONTGOMERY_SHIFT: u8 = 16;
+const MONTGOMERY_R: i32 = 1 << MONTGOMERY_SHIFT;
 
-#[cfg_attr(hax, hax_lib_macros::requires(value >= -106527 && value <= 104862))]
+#[cfg_attr(hax, hax_lib_macros::ensures(|result| result < 2u32.pow(MONTGOMERY_SHIFT as u32)))]
+#[inline(always)]
+fn get_montgomery_r_least_significant_bits(value: u32) -> u32 {
+    value & ((1 << MONTGOMERY_SHIFT) - 1)
+}
+
+const BARRETT_SHIFT: i64 = 26;
+const BARRETT_R: i64 = 1 << BARRETT_SHIFT;
+const BARRETT_MULTIPLIER: i64 = 20159; // floor((BARRETT_R / FIELD_MODULUS) + 0.5)
+
+#[cfg_attr(hax, hax_lib_macros::requires((i64::from(value) > -(BARRETT_R / 2) && i64::from(value) < BARRETT_R / 2)))]
 #[cfg_attr(hax, hax_lib_macros::ensures(|result| result > -FIELD_MODULUS && result < FIELD_MODULUS))]
 pub(crate) fn barrett_reduce(value: KyberFieldElement) -> KyberFieldElement {
-    let quotient = ((value * BARRETT_MULTIPLIER) + (BARRETT_R >> 1)) >> BARRETT_SHIFT;
+    let t = (i64::from(value) * BARRETT_MULTIPLIER) + (BARRETT_R >> 1);
+    let quotient = (t >> BARRETT_SHIFT) as i32;
 
     value - (quotient * FIELD_MODULUS)
 }
 
-const MONTGOMERY_SHIFT: i32 = 16;
-const MONTGOMERY_R: i32 = 1i32 << MONTGOMERY_SHIFT;
-const INVERSE_OF_MODULUS_MOD_R: i32 = -3327; // FIELD_MODULUS^{-1} mod MONTGOMERY_R
+const INVERSE_OF_MODULUS_MOD_R: u32 = 62209; // FIELD_MODULUS^{-1} mod MONTGOMERY_R
 
-#[cfg_attr(hax, hax_lib_macros::requires(value > -FIELD_MODULUS * MONTGOMERY_R / 2 && value < FIELD_MODULUS * MONTGOMERY_R / 2))]
-#[cfg_attr(hax, hax_lib_macros::ensures(|result| result > -FIELD_MODULUS && result < FIELD_MODULUS))]
+#[cfg_attr(hax, hax_lib_macros::requires(value >= -FIELD_MODULUS * MONTGOMERY_R && value <= FIELD_MODULUS * MONTGOMERY_R))]
+#[cfg_attr(hax, hax_lib_macros::ensures(|result| result >= -(3 * FIELD_MODULUS) / 2 && result <= (3 * FIELD_MODULUS) / 2))]
 pub(crate) fn montgomery_reduce(value: KyberFieldElement) -> KyberFieldElement {
-    let k = (value & (MONTGOMERY_R - 1)) * INVERSE_OF_MODULUS_MOD_R;
-    let k = (k & (MONTGOMERY_R - 1)) as i16;
+    // This forces hax to extract code for MONTGOMERY_R before it extracts code
```

Comment:
Please add an issue to anything that should get changed again, like this.
Maybe just file one issue and link everything to it.